File: golang/src/multijob/multijob.go
// Parse the Multijob command line interface.
// This allows you to implement Multijob tasks in Go.
//
//      args, err := multijob.ParseCommandLine(args, nil)
//      if err != nil {
//          ...
//      }
//
//      x, err := args.GetStr("x")
//      if err != nil {
//          ...
//      }
//
//      if err = args.NoFurtherArguments(); err != nil {
//          ...
//      }
package multijob

import "fmt"
import "strconv"

// JobArgvConfig lets you change the names of the special "--id" and "--rep"
// parameters.
type JobArgvConfig struct {
	JobIDKey        string // JobIDKey is the name of the "--id" argument.
	RepetitionIDKey string // RepetitionIDKey is the name of the "--rep" argument.
}

var defaultJobArgvConfig JobArgvConfig = JobArgvConfig{"--id", "--rep"}

// ParseCommandline turns the command line arguments (see "os.Args") into a
// queryable "Args" object.
//
// Argument "argv" contains the arguments to be parsed.
//
// Argument "config" will usually be "nil", but can be explicitly provided to
// override the "--id" or "--rep" special argument names.
func ParseCommandline(argv []string, config *JobArgvConfig) (args *Args, err error) {
	if config == nil {
		config = &defaultJobArgvConfig
	}

	separator := "--"

	specialArgs, normalArgs, err := separateArgvIntoSpecialAndNormalKVs(
		argv, separator)

	jobIDStr, ok := specialArgs[config.JobIDKey]
	if !ok {
		err = fmt.Errorf("multijob: special JobID argument %q required",
			config.JobIDKey)
		return
	}

	repetitionIDStr, ok := specialArgs[config.RepetitionIDKey]
	if !ok {
		err = fmt.Errorf("multijob: special RepetitionID argument %q required",
			config.RepetitionIDKey)
		return
	}

	delete(specialArgs, config.JobIDKey)
	delete(specialArgs, config.RepetitionIDKey)

	if len(specialArgs) > 0 {
		err = fmt.Errorf(
			"multijob: unknown special arguments before %q separator: ",
			separator,
			joinKeys(specialArgs, " "))
		return
	}

	jobID, err := strconv.Atoi(jobIDStr)
	if err != nil {
		err = fmt.Errorf("multijob: can't parse JobID %q: %s", jobIDStr, err.Error())
		return
	}

	repetitionID, err := strconv.Atoi(repetitionIDStr)
	if err != nil {
		err = fmt.Errorf("multijob: can't parse RepetitionID %q: %s", repetitionIDStr, err.Error())
	}

	args = &Args{
		JobID:        jobID,
		RepetitionID: repetitionID,
		args:         normalArgs,
		argWasUsed:   make(map[string]bool),
	}
	return
}

// Args represents the parsed arguments.
// You may retrieve arguments via the "Get*" methods, e.g. "GetStr()".
type Args struct {
	JobID        int
	RepetitionID int
	args         map[string]string
	argWasUsed   map[string]bool
}

func (args *Args) NoFurtherArguments() (err error) {
	unusedArgs := make([]string, 0, len(args.args))
	for k, _ := range args.args {
		if _, ok := args.argWasUsed[k]; !ok {
			unusedArgs = append(unusedArgs, k)
		}
	}

	if len(unusedArgs) > 0 {
		err = fmt.Errorf(
			"multijob: Unused arguments remain: %s",
			joinSortedQuotedItems(unusedArgs, ", "))
	}

	return
}

// GetStr retrieves a string value from the command line arguments.
func (args *Args) GetStr(key string) (value string, err error) {
	value, ok := args.args[key]
	if !ok {
		err = fmt.Errorf("multijob: no %q argument", key)
	}
	args.argWasUsed[key] = true
	return
}

File: cpp/multijob.cpp
#include "include/multijob.h"
#include "MultijobError.h"
#include "JoinedAndQuoted.h"
#include "conversion.h"

#include <string>
#include <iomanip>
#include <tuple>
#include <algorithm>

namespace
{
    auto split_arg(std::string const& arg)
        -> std::pair<std::string, std::string>
    {
        auto sep_pos = arg.find('=');

        if (sep_pos == std::string::npos)
            throw MULTIJOB_ERROR("can't split as argument: " << std::quoted(arg));

        auto key    = arg.substr(0, sep_pos);
        auto value  = arg.substr(sep_pos + 1);

        return {key, value};
    }

    using MapStrStr = std::unordered_map<std::string, std::string>;

    auto separate_argv(
            int argc,
            char const* const* argv,
            std::string const& sep)
        -> std::pair<MapStrStr, MapStrStr>
    {
        char const* const* end = &argv[argc];

        MapStrStr special_args;
        for (; argv != end; ++argv)
        {
            if (*argv == sep)
                break;

            std::string key;
            std::string value;
            std::tie(key, value) = split_arg(*argv);
            special_args[key] = std::move(value);
        }

        if (argv != end)
            ++argv;

        MapStrStr normal_args;
        for (; argv != end; ++argv)
        {
            std::string key;
            std::string value;
            std::tie(key, value) = split_arg(*argv);
            normal_args[key] = std::move(value);
        }

        return {std::move(special_args), std::move(normal_args)};
    }

  }

namespace multijob
{

auto parse_commandline(
        int argc,
        char const* const* argv,
        JobArgvConfig const& config)
    -> Args
{
    // skip argv[0], which is the executable name
    --argc;
    ++argv;

    MapStrStr special_args;
    MapStrStr normal_args;
    std::tie(special_args, normal_args) = separate_argv(argc, argv, "--");

    auto job_id_it = special_args.find(config.job_id_key);
    if (job_id_it == special_args.end())
    {
        throw MULTIJOB_ERROR(
                "special job_id argument " <<
                std::quoted(config.job_id_key) << " required");
    }
    auto job_id_str = job_id_it->second;
    special_args.erase(job_id_it);

    auto repetition_id_it = special_args.find(config.repetition_id_key);
    if (repetition_id_it == special_args.end())
    {
        throw MULTIJOB_ERROR(
                "special repetition_id argument " <<
                std::quoted(config.repetition_id_key) << " required");
    }
    auto repetition_id_str = repetition_id_it->second;
    special_args.erase(repetition_id_it);

    if (special_args.size())
    {
        std::vector<std::string> keys;
        for (auto const& kv : special_args)
            keys.emplace_back(kv.first);
        std::sort(keys.begin(), keys.end());

        throw MULTIJOB_ERROR(
                "unknown special arguments before " << std::quoted("--") <<
                " separator: " <<
                joined_and_quoted(", ", keys.begin(), keys.end()));
    }

    ID job_id = ID(convert_str_to_ul("job_id", job_id_str));
    ID repetition_id = ID(convert_str_to_ul("repetition_id", repetition_id_str));

    return {job_id, repetition_id, normal_args};
}

}
